### PR TITLE
Various cmake cleanups

### PR DESCRIPTION
**Pull request by rschu1ze**
More cmake simplification and cleanup while teaching me cmake in preparation for another task.

See the individual commits for details.

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

File: cmake/ccache.cmake [hunk 1]
```diff
@@ -1,56 +1,51 @@
...
-elseif (NOT CCACHE_FOUND AND NOT COMPILER_MATCHES_CCACHE)
-    message (${RECONFIGURE_MESSAGE_LEVEL} "Using ccache: No")
-endif ()
+execute_process(COMMAND ${CCACHE_EXECUTABLE} "-V" OUTPUT_VARIABLE CCACHE_VERSION)
+string(REGEX REPLACE "ccache version ([0-9\\.]+).*" "\\1" CCACHE_VERSION ${CCACHE_VERSION})
+
+set (CCACHE_MINIMUM_VERSION 3.3)
+
+if (CCACHE_VERSION VERSION_LESS_EQUAL ${CCACHE_MINIMUM_VERSION})
+    message(FATAL_ERROR "Using ccache: no (found ${CCACHE_EXECUTABLE} (version ${CCACHE_VERSION}), the minimum required version is ${CCACHE_MINIMUM_VERSION}")
```
Comment (azat) [inline cmake/ccache.cmake:31]:
```suggestion
    message(${RECONFIGURE_MESSAGE_LEVEL} "Using ccache: no (found ${CCACHE_EXECUTABLE} (version ${CCACHE_VERSION}), the minimum required version is ${CCACHE_MINIMUM_VERSION}")
    return()
```

Comment (rschu1ze) [reply cmake/ccache.cmake:31]: I wasn't exactly sure either but your proposal sounds good.

**Comment by alexey-milovidov** [comment]
Interesting, how did we miss `-Wextra-semi`?

File: cmake/ccache.cmake [hunk 2]
```diff
@@ -25,25 +25,27 @@ endif()
...
+endif()
+
+message(STATUS "Using ccache: ${CCACHE_EXECUTABLE} (version ${CCACHE_VERSION})")
+set(LAUNCHER ${CCACHE_EXECUTABLE})
+
+# debian (debhelpers) set SOURCE_DATE_EPOCH environment variable, that is
```
Comment (alexey-milovidov) [inline cmake/ccache.cmake:38]: Just in case, we don't use debhelpers anymore.

File: cmake/warnings.cmake [hunk 3]
```diff
@@ -9,93 +9,46 @@
...
-option (WEVERYTHING "Enable -Weverything option with some exceptions." ON)
-
 # Control maximum size of stack frames. It can be important if the code is run in fibers with small stack size.
 # Only in release build because debug has too large stack frames.
 if ((NOT CMAKE_BUILD_TYPE_UC STREQUAL "DEBUG") AND (NOT SANITIZE) AND (NOT CMAKE_CXX_COMPILER_ID MATCHES "AppleClang"))
     add_warning(frame-larger-than=65536)
 endif ()
 
 if (COMPILER_CLANG)
+    add_warning(everything)
```
Comment (alexey-milovidov) [inline cmake/warnings.cmake:19]: I'm actually very proud of our usage of `-Weverything`
let's keep the comment.

Comment (alexey-milovidov) [reply cmake/warnings.cmake:21]: Added a comment.

**Review by alexey-milovidov — APPROVED** [review]

File: cmake/ccache.cmake (hunk 2 above)
Comment (rschu1ze) [reply cmake/ccache.cmake]: Clarified the comment.

File: cmake/ccache.cmake [hunk 4]
```diff
@@ -1,56 +1,53 @@
...
+# environment variable SOURCE_DATE_EPOCH. This variable provides an alternative
+# to source-code embedded timestamps (__DATE__/__TIME__) and therefore helps with
+# reproducible builds (*). SOURCE_DATE_EPOCH is set automatically by the
+# distribution, e.g. Debian. Ccache 4.0 & 4.1 incorporate SOURCE_DATE_EPOCH into
+# the hash calculation regardless they contain timestamps or not. This invalidates
+# the cache whenever SOURCE_DATE_EPOCH changes. As a fix, ignore SOURCE_DATE_EPOCH.
+#
+# (*) https://reproducible-builds.org/specs/source-date-epoch/
+if (CCACHE_VERSION VERSION_GREATER_EQUAL "4.0" AND CCACHE_VERSION VERSION_LESS "4.2")
+    message(STATUS "Ignore SOURCE_DATE_EPOCH for ccache 4.1 / 4.2")
```
Comment (azat) [inline cmake/ccache.cmake:48]:
```suggestion
    message(STATUS "Ignore SOURCE_DATE_EPOCH for ccache 4.0 / 4.1")
```

Comment (rschu1ze) [reply cmake/ccache.cmake]: Thanks (guess I need to get another coffee  :D).

**Comment by rschu1ze** [comment]
> Interesting, how did we miss `-Wextra-semi`?

As far as I see, it is enabled by "-wc++98-compat" which was (rightfully) disabled in the WEVERYTHING path. Will check independently if the warning can be enabled using some other check. But in the end it is cosmetics only.

**Pull request merged**